Use the Open edX named releases to conditionally import platform objects

johnbaldwin opened the pull request:
This commit exists to address the problem that a RuntimeError can be
raised instead of an ImportError when importing objects from
ed-platform.

The options to fix are check for ImportError AND RuntimeError. However,
we opt (at least for now) to check the Open edX named release to
conditionally import objects when the namespace has changed between
named releases.

Here is an example of the error when only ImportError is handled:
```
RuntimeError: Conflicting 'studentmodule' models in application 'courseware': <class 'courseware.models.StudentModule'> and <class 'lms.djangoapps.courseware.models.StudentModule'>.
```

NOTE: This may break Juniper compat as it was not tested in a live
Juniper devstack

NOTE: See @OmarIthawi 's comments and this issue: https://github.com/appsembler/figures/issues/290

OmarIthawi commented on figures/compat.py:
```diff
@@ -11,46 +11,35 @@
 
-try:
+if RELEASE_LINE == 'juniper':
     from lms.djangoapps.courseware.courses import get_course_by_id  # noqa pylint: disable=unused-import,import-error
```
Should work fine until Koa.

OmarIthawi commented on figures/compat.py:
```diff
@@ -11,46 +11,35 @@
     from courseware.models import StudentModule  # noqa pylint: disable=unused-import,import-error
 
-try:
+if RELEASE_LINE == 'juniper':
```
Should work fine in Koa.

```suggestion
if RELEASE_LINE in ['juniper', 'koa']:
```

OmarIthawi commented on figures/compat.py:
```diff
@@ -11,46 +11,35 @@
 
-
-try:
+if RELEASE_LINE == 'juniper':
```
```suggestion
if RELEASE_LINE in ['juniper', 'koa']:
```

```
# koa
lms/djangoapps/instructor_analytics/basic.py:from lms.djangoapps.courseware.models import StudentModule
```

OmarIthawi commented on figures/compat.py:
```diff
@@ -11,46 +11,35 @@
-try:
-    from opaque_keys.edx.django.models import CourseKeyField  # noqa pylint: disable=unused-import,import-error
-except ImportError:
+if RELEASE_LINE == 'ginkgo':
```
This is future-proof. Great!

OmarIthawi commented on figures/compat.py:
```diff
@@ -11,46 +11,35 @@
     from lms.djangoapps.courseware.courses import get_course_by_id  # noqa pylint: disable=unused-import,import-error
-except ImportError:
-    # Backward compatibily for pre-Juniper releases
+else:  # Backward compatibily for pre-Juniper releases
```
This import is not future proof. 

Those imports would work but are not future-proof. To fix it needs to be restructured in the following form from the oldest to the most recent release:

```python
if ginkgo or hawthorn:
  import # old import
else:   # all future releases
  import  # future proof
```

OmarIthawi reviewed (approved):
I forgot to specify that all my previous comments where non-blockers. LGTM. Please feel free to merge. Also, thanks for making the issue #290.

johnbaldwin commented:
@OmarIthawi Thanks and I'm merging this in. Take another look. I believe I addressed your comments. Only thing I think needs to be addressed is getting test coverage for `figures.compat.course_grades`. See my git commit message for details

OmarIthawi commented on figures/compat.py:
```diff
@@ -11,46 +11,45 @@
 except ImportError:
-    # we are pre-ginkgo
-    RELEASE_LINE = None
+    raise UnsuportedOpenedXRelease(
```
That's good!

OmarIthawi commented:
Thanks @johnbaldwin! I think this is pretty good and probably don't need the additional coverage.
